src/librustc/session/config.rs

```diff
@@ -1285,6 +1283,95 @@ pub fn parse_crate_types_from_list(list_list: Vec<String>) -> Result<Vec<CrateTy
     return Ok(crate_types);
 }
 
+pub struct NightlyOptions {
+    is_nightly_build: bool,
+    is_unstable_enabled: bool,
+    flags: Vec<RustcOptGroup>,
+}
+
+impl NightlyOptions {
+    pub fn get(matches: &getopts::Matches) -> NightlyOptions {
+        let nightly_build = get_unstable_features_setting();
+        NightlyOptions {
+            is_nightly_build: match nightly_build {
+                                  UnstableFeatures::Allow | UnstableFeatures::Cheat => true,
+                                  _ => false,
+                              },
+            is_unstable_enabled: matches.opt_strs("Z").iter().any(|x| *x == "unstable-options"),
+            flags: rustc_optgroups(),
+        }
+    }
+
+    pub fn is_unstable(&self, option_name: &str) -> bool {
+        self.flags.iter().any(|x| {
+            x.opt_group.long_name == option_name || x.opt_group.short_name == option_name
+        })
+    }
+
+    pub fn is_unstable_enabled(&self) -> bool {
+        self.is_nightly_build() && self.is_unstable_enabled
+    }
+
+    pub fn is_nightly_build(&self) -> bool {
+        self.is_nightly_build
+    }
+
+    pub fn check_enabled(&self, enable_flag: &str, flag: &str) {
+        let e = if !self.is_nightly_build() {
+            format!("the option `{}` is only accepted on the nightly compiler", flag)
+        } else if !self.is_unstable_enabled() {
+            format!("the `-Z {}` flag must also be passed to enable the flag `{}`",
+                    enable_flag, flag)
+        } else {
+            return
+        };
+        early_error(ErrorOutputType::default(), &e)
+    }
+
+    pub fn is_enabled(&self, matches: &getopts::Matches, option: &str) -> bool {
```

Comment:
Perhaps this structure could just be exposed as a function? Most of these `pub` methods aren't `pub` to be called elsewhere, and it's good to keep the API surface area small if possible